### PR TITLE
Fix class name collisions

### PR DESCRIPTION
**Pull request by staabm**
similar to https://github.com/phpstan/phpstan-src/commit/95cdbe577513286c36dcf513fe76f269e8a32125

### CODE REVIEW

File: .github/workflows/code_analysis.yaml [hunk 1]
```diff
@@ -83,7 +88,7 @@ jobs:
             -
                 uses: shivammathur/setup-php@v2
                 with:
-                    php-version: 8.1
+                    php-version: 8.2
```
Comment (staabm) [inline .github/workflows/code_analysis.yaml]: is it fine, that I have raised the php version for this analysis jobs?

[without, the job failed](https://github.com/rectorphp/rector-src/actions/runs/6186775775/job/16795178900?pr=5018)

**Comment by staabm** [comment]
I am not sure, whether the test actually started working because of my changes, or whether we see a valid error?

**Comment by samsonasik** [comment]
not natively property based type should be skipped, as its source use `@var`

https://github.com/rectorphp/rector-src/blob/2b177e80e2630c97653aa9d57ac87acbc1433524/rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector/Source/NotNativelyTypedDataObject.php#L9-L13

**Comment by staabm** [comment]
> not natively property based type should be skipped, as its source use `@var`
> 
> https://github.com/rectorphp/rector-src/blob/2b177e80e2630c97653aa9d57ac87acbc1433524/rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector/Source/NotNativelyTypedDataObject.php#L9-L13

thanks, fixed in [8dcb43e](https://github.com/rectorphp/rector-src/pull/5018/commits/8dcb43ed2c49c5333eb7935c293c0dcc280ec41e)

File: rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/TrustedClassMethodPropertyTypeInferer.php [hunk 2]
```diff
@@ -73,7 +73,7 @@ public function inferProperty(Class_ $class, Property $property, ClassMethod $cl
 
         $resolvedTypes = [];
         foreach ($assignedExprs as $assignedExpr) {
-            $resolvedTypes[] = $this->nodeTypeResolver->getType($assignedExpr);
+            $resolvedTypes[] = $this->nodeTypeResolver->getNativeType($assignedExpr);
```
Comment (staabm) [inline rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/TrustedClassMethodPropertyTypeInferer.php]: thats a actual bugfix for https://github.com/rectorphp/rector-src/pull/5018#issuecomment-1719578352

File: rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromStrictConstructorRector/Fixture/public_property_doc.php.inc [hunk 3]
```diff
@@ -25,7 +25,7 @@ use stdClass;
 final class PublicPropertyDoc
 {
     /**
-     * @var stdClass
+     * @var \stdClass
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromStrictConstructorRector/Fixture/public_property_doc.php.inc]: not sure, why rector is now reporting a FQCN, but I think its fine

File: rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector/Source/NotNativelyTypedDataObject.php [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 declare(strict_types=1);
 
-namespace Rector\Tests\TypeDeclaration\Rector\Property\TypedPropertyFromStrictConstructorRector\Source;
+namespace Rector\Tests\TypeDeclaration\Rector\Property\TypedPropertyFromAssignsRector\Source;
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector/Source/NotNativelyTypedDataObject.php]: the name-collision tool detected this problem, and from here on the journey began which lead to a [actual bug beeing fixed](https://github.com/rectorphp/rector-src/pull/5018#issuecomment-1719578352)

File: .github/workflows/code_analysis.yaml (hunk 1 above)
Comment (TomasVotruba) [reply .github/workflows/code_analysis.yaml]: I want to strick with PHP 8.1 now, as I run it everywhere and Rector works well on it.

File: rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/TrustedClassMethodPropertyTypeInferer.php (hunk 2 above)
Comment (samsonasik) [reply rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/TrustedClassMethodPropertyTypeInferer.php]: Could you cherry pick this patch in separate PR? Thank you.

Comment (staabm) [reply rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/TrustedClassMethodPropertyTypeInferer.php]: done https://github.com/rectorphp/rector-src/pull/5020

File: .github/workflows/code_analysis.yaml (hunk 1 above)
Comment (staabm) [reply .github/workflows/code_analysis.yaml]: done, added the failling case to the exclude path of the collision tool

File: build/collision-detector.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+    "scanPaths": ["../build", "../packages", "../packages-tests", "../rules", "../rules-tests", "../src", "../tests", "../tests-paths", "../utils", "../utils-tests"],
+    "excludePaths": [
+        "../rules-tests/Php82/Rector/Class_/ReadOnlyClassRector/Source/ParentAlreadyReadonly.php"
```
Comment (staabm) [inline build/collision-detector.json]: skipped because it requires php 8.2

**Comment by TomasVotruba** [comment]
Thanks for the fixture improvement + fix.

I don't think it's worth to have extra tool for base name check. Especially when `/build` is run standalone and reports false positive.
It might be good idea now, but we'll have to maintain and update it ever since.

I'd prefer to have this within PHPStan, as its scope. E.g. improve https://github.com/symplify/phpstan-rules/blob/main/docs/rules_overview.md#noduplicatedshortclassnamerule to catch what was missed here.

**Comment by staabm** [comment]
here you are

**Comment by staabm** [comment]
> I'd prefer to have this within PHPStan, as its scope. E.g. improve [symplify/phpstan-rules@`main`/docs/rules_overview.md#noduplicatedshortclassnamerule](https://github.com/symplify/phpstan-rules/blob/main/docs/rules_overview.md?rgh-link-date=2023-09-14T17%3A26%3A48Z#noduplicatedshortclassnamerule) to catch what was missed here.

I don't think this rule can catch the problem we had in https://github.com/rectorphp/rector-src/pull/5020/files#diff-29b1a9f53e1c99445ccc574568939f257862caa8b9adc2529b9aaa189d6b76cbR5
(which was a wrong PSR namespace)

File: build/target-repository/e2e/parse-php7-code/composer.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
     "autoload": {
         "psr-4": {
-            "Foo\\": "src"
+            "Foo7\\": "src"
```
Comment (samsonasik) [inline build/target-repository/e2e/parse-php7-code/composer.json:4]: e2e test is run independently, so duplicated namespace is fine. it executed inside of per-e2e sub dir:

https://github.com/rectorphp/rector/blob/7c6e8362953e79ed92b439fd823595eb3c5ec09f/.github/workflows/e2e.yaml#L20-L40

Comment (staabm) [reply build/target-repository/e2e/parse-php7-code/composer.json:4]: I am fine with whatever you guys decide - no strong opinion on my end

**Comment by TomasVotruba** [comment]
> I don't think this rule can catch the problem we had in [#5020 (files)](https://github.com/rectorphp/rector-src/pull/5020/files#diff-29b1a9f53e1c99445ccc574568939f257862caa8b9adc2529b9aaa189d6b76cbR5)
(which was a wrong PSR namespace)

That's exactly what I expect from PHPStan. It possible the `*/Source/*` files are skipped by PHPStan. We could enable the class-exist for the `/Source/` directory to validate this.

**Comment by TomasVotruba** [comment]
@staabm I'll use Occams razor and stick withl PHPStan for now. Thanks for fixes in the other PR :+1:

**Pull request closed without merging**